index/segment/mem/segment_simple.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/uber-go/atomic"
+)
+
+// TODO(prateek): investigate impact of native heap
+type simpleSegment struct {
+	opts     Options
+	id       segment.ID
+	docIDGen *atomic.Uint32
+
+	// internal docID -> document
+	documentsLock sync.RWMutex
+	documents     map[segment.DocID]document // TODO(prateek): measure perf impact of slice v map here
+
+	// field (Name+Value) -> postingsManagerOffset
+	termsDict termsDictionary
+
+	searcher searcher
+	// TODO(prateek): add a delete documents bitmap to optimise fetch
+}
+
+type document struct {
+	doc.Document
+	docID      segment.DocID
+	tombstoned bool
+}
+
+// New returns a new in-memory index.
+func New(id segment.ID, opts Options) (Segment, error) {
+	seg := &simpleSegment{
+		opts:     opts,
+		id:       id,
+		docIDGen: atomic.NewUint32(0),
+
+		documents: make(map[segment.DocID]document, opts.InitialCapacity()),
+		termsDict: newSimpleTermsDictionary(opts),
+	}
+
+	searcher := newSequentialSearcher(seg, differenceNegationFn, opts.PostingsListPool())
+	seg.searcher = searcher
+	return seg, nil
+}
+
+// TODO(prateek): consider copy semantics for input data, esp if we can store it on the native heap
+func (i *simpleSegment) Insert(d doc.Document) error {
+	return i.insertDocument(document{
+		Document: d,
+		docID:    segment.DocID(i.docIDGen.Inc()),
+	})
+}
+
+func (i *simpleSegment) insertDocument(doc document) error {
+	// insert document into master doc id -> doc map
+	i.documentsLock.Lock()
+	i.documents[doc.docID] = doc
```

Comment:
We'll have a segment per shard yeah?  Otherwise a simple top level mutex seems like it might cause a lot of contention.

Reply:
Nah I'm thinking of keeping segment and shard independent. It'll be too expensive to keep a segment per shard on the read side – every fetch would have to hit every segment in each shard; i.e. thousands of map lookups. 

My plan for contention was to have an insert channel at the top-level index.

Reply:
If the indices are independent of the shards though how do we plan on splitting indices when nodes join a cluster and shards are assigned to different hosts? We won't be able to just move the index along with the shard like we would be able to do with an index per shard approach.

Reply:
Two parts: 
- For the node with the leaving shard, we'd expect the shard to issue a `Delete()` for all it's IDs so that the index doesn't have any zombie docs
- For the node with the joining shard, it'd bootstrap the same as the normal process after bootstrap and get the appropriate Insert() called.